tests/federation_room_send_test.go

```diff
@@ -22,54 +21,47 @@ func TestOutboundFederationSend(t *testing.T) {
 	deployment := Deploy(t, "federation_send", b.BlueprintAlice)
 	defer deployment.Destroy(t)
 
+	waiter := NewWaiter()
+	wantEventType := "m.room.message"
+
+	// create a remote homeserver
 	srv := federation.NewServer(t, deployment,
 		federation.HandleKeyRequests(),
 		federation.HandleMakeSendJoinRequests(),
+		federation.HandleTransactionRequests(
+			// listen for PDU events in transactions
+			func(ev gomatrixserverlib.Event) {
+				defer waiter.Finish()
+
+				if ev.Type() != wantEventType {
+					t.Errorf("Wrong event type, got %s want %s", ev.Type(), wantEventType)
+				}
+			},
+			nil,
+		),
 	)
 	cancel := srv.Listen()
 	defer cancel()
 
+	// the remote homeserver creates a public room
 	ver := gomatrixserverlib.RoomVersionV5
 	charlie := srv.UserID("charlie")
 	serverRoom := srv.MustMakeRoom(t, ver, federation.InitialRoomEvents(ver, charlie))
 	roomAlias := srv.MakeAliasMapping("flibble", serverRoom.RoomID)
 
-	// join the room
+	// the local homeserver joins the room
 	alice := deployment.Client(t, "hs1", "@alice:hs1")
-	alice.JoinRoom(t, roomAlias, nil)
-
-	wantEventType := "m.room.message"
-
-	waiter := NewWaiter()
-
-	// TODO: Have a nicer api shape than just http.Handler
-	srv.Mux().Handle("/_matrix/federation/v1/send/{txnID}", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		defer waiter.Finish()
-		var body gomatrixserverlib.Transaction
-		if err := json.NewDecoder(req.Body).Decode(&body); err != nil {
-			t.Errorf("failed to decode /send body: %s", err)
-			w.WriteHeader(500)
-			return
-		}
-		if len(body.PDUs) != 1 {
-			t.Fatalf("got %d pdus, want %d", len(body.PDUs), 1)
-		}
```

Comment:
We're losing this check in this change, though it's not as simple to perform this check as we only get the event instead of the transaction.

I'm wondering whether we should also pass the entire transaction to the callback in case it wants to check other parts of it?

Reply:
I'd leave it out, if people care about # events in a txn then they shouldn't use this helper.